coordinator.lic

```diff
@@ -0,0 +1,478 @@
+      end
+      send(value)
+    else
+      raise ArgumentError, "Unknown predicate type: #{key}"
+    end
+  end
+
+  def prepare_tasks
+    update_box_count
+    @local_timers = {}
+    @cleanup_tasks.each { |task| task[:marked] = !task[:skip_first_run] }
+  end
+
+  def terminal_predicates(task)
+    predicates = []
+    (predicates << task[:start_on]) if task[:start_on]
+    (predicates << task[:stop_on]) if task[:stop_on]
+    predicates.map { |pred| terminal_predicates_helper(pred) }.flatten
+  end
+
+  def terminal_predicates_helper(predicate)
+    return [] if predicate.nil? || predicate.empty?
+    unless predicate.size == 1
+      raise ArgumentError, "Predicate with multiple keys: #{predicate}"
+    end
+
+    key = predicate.keys.first
+    value = predicate[key]
+    case key
+    when 'or', 'and'
+      value.map { |member| terminal_predicates_helper(member) }.flatten
+    else
+      [predicate]
+    end    
+  end
+
+  def get_timers(task)
+    terminal_predicates(task).select { |pred| pred.keys.first == 'timer' }
+      .map { |pred| pred['timer'] }
+  end
+
+  def update_predicates_for_completed_task(task)
+    if task[:type] == :cleanup
+      task[:marked] = false
+    elsif task[:type] == :hunting && !task[:no_cleanup]
+      @cleanup_tasks.each { |task| task[:marked] = true }
+    end
+    get_timers(task).each { |timer| reset_timer(timer) }
+  end
+  
+  def reset_timer(timer)
+    reset_time = Time.now + timer['time']
+    if timer['global']
+      echo("Resetting global timer: #{timer}") if @debug      
+      UserVars.coordinator_timers[timer['key']] = reset_time
+    else
+      echo("Resetting local timer: #{timer}") if @debug      
+      @local_timers[timer['key']] = reset_time
+    end
+  end
+
+  def get_time(timer)
+    if timer['global']
+      UserVars.coordinator_timers[timer['key']] || (Time.now - 1)
+    else
+      @local_timers[timer['key']] || (Time.now - 1)
+    end    
+  end
+  
+  def skill_less_than?(skill, target)
+    echo("Skill level is #{DRSkill.getxp(skill)}") if @debug
+    DRSkill.getxp(skill) < target
+  end
+
+  def skill_greater_than?(skill, target)
+    echo("Skill level is #{DRSkill.getxp(skill)}") if @debug
+    DRSkill.getxp(skill) > target
+  end
+
+  def update_box_count
+    @num_boxes = (get_boxes(@settings.picking_box_source) || []).size
+    @num_pet_boxes = (get_boxes(@settings.picking_pet_box_source) || []).size
+  end
+
+  def run_town_task(task)
+    echo("Running town task #{task}") if @debug
+    safe_room = @settings.safe_room
+    pause_script('sanowret-crystal') if task[:pause_sanowret]
+    if task[:walk_to]
+      case task[:walk_to]
+      when 'safe_room'
+        walk_to(safe_room)
+      else
+        walk_to(task[:walk_to])
+      end
+    end
+    if task[:buff]
+      wait_for_script_to_complete('buff', [task[:buff]])
+    end
+    if task[:town] && DRStats.necromancer?
+      wait_for_script_to_complete('release-necro')
```

Comment:
💡 This script isn't in this repository. What will to a user that triggers this case?